reconcile/openshift_base.py

```diff
@@ -556,18 +570,441 @@ def check_unused_resource_types(ri):
+        c_item_sha256 = annotations["qontract.sha256sum"]
+    except KeyError:
+        logging.info("Current object QR hash is missing -> Apply")
+        return False
+
+    # Original object does not match Desired -> Apply
+    # Current object is not recalculated!
+    # d_item_sha256 = OR.calculate_sha256sum(OR.serialize(d_item.body))
+    # if c_item_sha256 != d_item_sha256:
+    if c_item_sha256 != d_item.sha256sum():
+        logging.info("Original and Desired objects hash differs -> Apply")
+        return False
+
+    # If there are differences between current and desired -> Apply
+    # The patch only detects changes with attributes defined in the desired state.
+    # Values in the current state added by operators or other actors are not taken
+    # into account
+
+    current = normalize_object(c_item)
+    desired = normalize_object(d_item)
+
+    patch = jsonpatch.JsonPatch.from_diff(current.body, desired.body)
+    for item in patch.patch:
+        if item["op"] == "add" or item["op"] == "replace":
+            # Add or Replace from DESIRED Over CURRENT
+            logging.info("Desired and Current objects differ -> Apply")
+            logging.info(patch.patch)
+            return False
+
+    return True
+
+
+@dataclass
+class ApplyOptions:
+    dry_run: bool
+    no_dry_run_skip_compare: bool
+    wait_for_namespace: bool
+    recycle_pods: bool
+    take_over: bool
+    override_enable_deletion: Optional[bool]
+    caller: Optional[str]
+    all_callers: Optional[Sequence[str]]
+    privileged: Optional[bool]
+    enable_deletion: Optional[bool]
+
+
+def handle_new_resources(
+    oc_map: OC_Map,
+    ri: ResourceInventory,
+    new_resources: Mapping[Any, Any],
+    cluster: str,
+    namespace: str,
+    resource_type: str,
+    data: Mapping[Any, Any],
+    options: ApplyOptions,
+) -> list[dict[str, Any]]:
+    actions: list[dict[str, Any]] = []
+
+    for name, desired in new_resources.items():
+        options.privileged = data["use_admin_token"].get(name, False)
+        action = {
+            "action": ACTION_APPLIED,
+            "cluster": cluster,
+            "namespace": namespace,
+            "kind": resource_type,
+            "name": name,
+            "privileged": options.privileged,
+        }
+        actions.append(action)
+        apply_action(
+            oc_map=oc_map,
+            ri=ri,
+            cluster=cluster,
+            namespace=namespace,
+            resource_type=resource_type,
+            resource=desired,
+            options=options,
+        )
+
+    return actions
+
+
+def handle_modified_resources(
+    oc_map: OC_Map,
+    ri: ResourceInventory,
+    modified_resources: Mapping[Any, Any],
+    cluster: str,
+    namespace: str,
+    resource_type: str,
+    data: Mapping[Any, Any],
+    options: ApplyOptions,
+) -> list[dict[str, Any]]:
+    actions: list[dict[str, Any]] = []
+
+    for name, dp in modified_resources.items():
+        b_apply = True
```

Comment:
what's `b_apply`? Is that more like should apply? Maybe extract filter logic out to make it clear:

```python
def should_apply(current, options, ri, cluster, name, namespace, resource_type):
    if not options.dry_run and options.no_dry_run_skip_compare:
        msg = (
            f"[{cluster}/{namespace}] skipping compare of resource"
            f"'{resource_type}/{name}'"
        )
        logging.debug(msg)
        return True
    if (
        not options.take_over
        and options.caller
        and current.caller != options.caller
        and options.all_callers
        and current.caller in options.all_callers
    ):
        ri.register_error()
        logging.error(
            f"[{cluster}/{namespace}] resource '{resource_type}/{name}' present and managed by another caller: {current.caller}"
        )
        return False
    logging.debug("CURRENT: " + OR.serialize(OR.canonicalize(current.body)))
    return True
```

Then it can be used like

```python
        if should_apply(current, options, ri, cluster, name, namespace, resource_type):
```

some params just for logging purpose, can also define that method inside `handle_modified_resources` to use closure to reduce some params.
```

Reply:
makes sense. PTAL